File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.py
class Solution:
    def feasible(self, weights: List[int], capacity: int, days: int) -> bool:
        daysNeeded, currentLoad = 1, 0
        for weight in weights:
            currentLoad += weight
            if currentLoad > capacity:
                daysNeeded += 1
                currentLoad = weight
        
        return daysNeeded <= days
        
        
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        totalLoad, maxLoad = 0, 0
        for weight in weights:
            totalLoad += weight
            maxLoad = max(maxLoad, weight)
            
        l, r = maxLoad, totalLoad
        
        while l < r:
            mid = (l + r) >> 1
            if self.feasible(weights, mid, days):
                r = mid
            else:
                l = mid + 1
        
        return l

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
    bool feasible(vector<int>& weights, int capacity, int days){
        int daysNeeded = 1, currentLoad = 0;
        
        for (int weight : weights){
            currentLoad += weight;
            if (currentLoad > capacity){
                currentLoad = weight;
                ++daysNeeded;
            }
        }
        return daysNeeded <= days;
    }
    
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int maxLoad = 0, totalLoad = 0;
        for (int weight : weights){
            totalLoad += weight;
            maxLoad = max(maxLoad, weight);
        }
        
        int l = maxLoad, r = totalLoad, mid;
        
        while (l < r){
            mid = (l + (r - l) / 2);
            if (feasible(weights, mid, days)) r = mid;
            else l = mid + 1;
        }
        
        return l;
    }
};